### PR TITLE
Add integration tests for various go mod corner cases.

### PR DESCRIPTION
**Pull request by Cynocracy**
"Fixes" https://github.com/google/ko/issues/171 (where I was confused by the behavior of modules) by ensuring ko behaves similarly to go build in various edge cases.

### CODE REVIEW

File: hack/tools.go [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package hack
+
+import (
+  "github.com/go-training/helloworld"
```
Comment (jonjohnsonjr) [inline hack/tools.go:4]: From travis:

> hack/tools.go:4:3: import "github.com/go-training/helloworld" is a program, not an importable package

I think you need a `_` before the import (and a `gofmt`).

Comment (Cynocracy) [reply hack/tools.go:4]: Thanks :)

Re: whether we should keep this behavior; totally agree, ideally we would replicate the behavior of go build exactly, which might be a moving target.

**Comment by Cynocracy** [comment]
Oh, because it's not actually obvious: I wrote the test cases before checking behavior :) So at least *I'm* convinced this behavior is correct.

**Comment by jonjohnsonjr** [comment]
Same error :/

I think you need a build tag that doesn't get used so that it doesn't try to build it, we did something similar in go-containerregistry: https://github.com/google/go-containerregistry/pull/599/files#diff-39bad9c50a0f30bd6c067b83431bec76R1

**Comment by Cynocracy** [comment]
Ah, I missed that I didn't have the tools build tag, yeah that's definitely necessary.

**Comment by Cynocracy** [comment]
Any suggestions on what to set for KO_DOCKER_REPO ?

**Comment by jonjohnsonjr** [comment]
> Any suggestions on what to set for KO_DOCKER_REPO ?

I think the current integration tests us the `--local` flag to just write it into the docker daemon, which doesn't require KO_DOCKER_REPO to be set, if you wanna do that.

**Comment by Cynocracy** [comment]
Ah, perfect, that works for me, thank you :)

**Comment by Cynocracy** [comment]
Ready! I think :)
